framework/finalizer.go

```diff
@@ -23,35 +23,61 @@ type patchSpec struct {
 }
 
 func (f *Framework) addFinalizer(obj interface{}) (stopReconciliation bool, err error) {
-	restClient := f.k8sClient.CoreV1().RESTClient()
-	patch, path, result, err := createAddFinalizerPatch(obj, f.name)
+	// We get the accessor of the object which we got passed from the framework.
+	accessor, err := meta.Accessor(obj)
 	if err != nil {
 		return false, microerror.Mask(err)
 	}
-	if patch == nil {
-		return result, err
-	}
-	p, err := json.Marshal(patch)
-	if err != nil {
-		return false, microerror.Mask(err)
+	finalizerName := getFinalizerName(f.name)
+	// We check if the object has a finalizer here, to avoid unnecessary calls to
+	// the k8s api.
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
```

Comment:
I made this check more defensive, this way we don't hit k8s at all if there is already a finalizer.